modules/core/02-client/abci_test.go

```diff
@@ -74,3 +76,55 @@ func (suite *ClientTestSuite) TestBeginBlockerConsensusState() {
 	suite.Require().NoError(err)
 	suite.Require().Equal(bz, consState)
 }
+
+func (suite *ClientTestSuite) TestBeginBlockerUpgradeEvents() {
+	plan := &upgradetypes.Plan{
+		Name:   "test",
+		Height: suite.chainA.GetContext().BlockHeight() + 1,
+	}
+	// set upgrade plan in the upgrade store
+	store := suite.chainA.GetContext().KVStore(suite.chainA.GetSimApp().GetKey(upgradetypes.StoreKey))
+	bz := suite.chainA.App.AppCodec().MustMarshal(plan)
+	store.Set(upgradetypes.PlanKey(), bz)
+
+	nextValsHash := []byte("nextValsHash")
+	newCtx := suite.chainA.GetContext().WithBlockHeader(tmproto.Header{
+		Height:             suite.chainA.GetContext().BlockHeight(),
+		NextValidatorsHash: nextValsHash,
+	})
+
+	err := suite.chainA.GetSimApp().UpgradeKeeper.SetUpgradedClient(newCtx, plan.Height, []byte("client state"))
+	suite.Require().NoError(err)
+
+	cacheCtx, writeCache := suite.chainA.GetContext().CacheContext()
+
+	client.BeginBlocker(cacheCtx, suite.chainA.App.GetIBCKeeper().ClientKeeper)
+	writeCache()
+
+	suite.requireContainsEvent(cacheCtx.EventManager().Events(), types.EventTypeUpgradeChain, true)
```

Comment:
just as a note, the caching of the context should be unnecessary in this situation. You should be able to get the event via `suite.chainA.GetContext()` after calling begin blocker on that context